aesd-char-driver/aesd-circular-buffer.c

```diff
@@ -0,0 +1,137 @@
+ *      buffptr member corresponding to char_offset.  This value is only set when a matching char_offset is found
+ *      in aesd_buffer.
+ * @return the struct aesd_buffer_entry structure representing the position described by char_offset, or
+ * NULL if this position is not available in the buffer (not enough data is written).
+ */
+struct aesd_buffer_entry *aesd_circular_buffer_find_entry_offset_for_fpos(struct aesd_circular_buffer *buffer,
+            size_t char_offset, size_t *entry_offset_byte_rtn )
+{
+    /* NULL check */
+    if((!buffer) || (!entry_offset_byte_rtn))
+    {
+        return NULL;
+    }
+
+    /* total bytes store the overall size of all the entries.
+     * tmp_offset decremental offset to find the offset within an entry */
+    size_t total_bytes = 0, tmp_offset = char_offset;
+
+    /* read offset points to oldest buffer in the queue*/
+    uint8_t read_offset = buffer->out_offs, i = 0;
+
+    /* Iterate through the CB queue and find the required entry and offset */
+    for(i=0; i<AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED; i++)
+    {
+        total_bytes += buffer->entry[read_offset].size;
+
+        /* If char offset found in given entry, break. */
+        if(char_offset < total_bytes)
+        {
+            break;
+        }
+
+        /* Decrement temp offset by each entry's size */
+        tmp_offset -= buffer->entry[read_offset].size;
+
+        /* if read overflows, reset to 0 */
+        read_offset++;
+        if(read_offset == AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED)
+        {
+            read_offset = 0;
+        }
+
+    }
+    /* Check if total_bytes are valid and char_offset is in bound */
+    if((total_bytes < 0) || (char_offset >= total_bytes))
+    {
+        *entry_offset_byte_rtn = 0;
+        return NULL;
+    }
+
+    /* Actual entry number based on the read offset */
+    *entry_offset_byte_rtn = tmp_offset;
+
+    /* Return the actual buffer entry */
+    return &(buffer->entry[read_offset]);
+}
+
+/**
+* Adds entry @param add_entry to @param buffer in the location specified in buffer->in_offs.
+* If the buffer was already full, overwrites the oldest entry and advances buffer->out_offs to the
+* new start location.
+* Any necessary locking must be handled by the caller
+* Any memory referenced in @param add_entry must be allocated by and/or must have a lifetime managed by the caller.
+*/
+void aesd_circular_buffer_add_entry(struct aesd_circular_buffer *buffer, const struct aesd_buffer_entry *add_entry)
+{
+    /* NULL check for pointers */
+    if((!buffer) || (!add_entry))
+    {
+        return;
+    }
+
+    /* Populate the entry buffer irrespective of buffer is full or not */
+    buffer->entry[buffer->in_offs].buffptr = add_entry->buffptr;
+    buffer->entry[buffer->in_offs++].size = add_entry->size;
+    if(buffer->in_offs == AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED)
```

Comment:
@dash6424 not absolutely required but I'd probably use `>=` here to handle the case that the input offset was corrupt and ensure we